### ISSUE TITLE
Feature request: reserving memory buffer for workload on a cluster node

### ISSUE THREAD

**Issue by syurevich** [kubernetes/kubernetes#122789]
### What would you like to be added?

We would like to be able to reserve some amount of memory (`buffer`) for our workload on a cluster node. 
Scheduler should not over-subscribe a `New Allocatable` = [Allocatable](https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/#node-allocatable) - `buffer`.
[Node-pressure eviction](https://kubernetes.io/docs/concepts/scheduling-eviction/node-pressure-eviction/) should still use an “old” `Allocatable` for evicting pods.

### Why is this needed?

This memory buffer would allow us to withstand memory consumption spikes and avoid OOMKilled errors.
This memory buffer would be available for any pod from our workload if this pod happens to temporarily need significantly more memory than configured in a memory request setting.
Such memory buffer would be especially helpful if [Memory QoS](https://kubernetes.io/blog/2021/11/26/qos-memory-resources/) feature gets enabled. 

We have very dynamic workload. We run a service which executes customer’s builds. For every new build our service creates a new pod that lives for a short time.
Memory consumption of such a pod can vary. In some cases it can try to consume significantly more memory than configured in a memory request setting of the pod.
If a node is fully packed with pods, it can lead to OOMKilled error and pod eviction.

We can only roughly guess possible memory consumption value because it depends on build type and build configuration provided by a customer.
Therefore, memory request settings which we configure cannot be precise.
Currently, we need to configure quite high memory request setting to minimize a likelihood of OOMKilled errors. 
This is very cost inefficient because we waste a lot memory on a node.

**Comment by kannon92** [comment]
/sig node

This is a popular ask. Inplace pod feature and swap both call this out as a use case to support.

Have you looked into either of these for this?

**Comment by syurevich** [comment]
[in-place-update](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources) is a valuable feature but it does not directly address a problem of memory pressure on a node. Probably, the [Future Enhancement](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources#future-enhancements) could help. I mean this:  "Kubelet (or Scheduler) evicts lower priority Pods from Node to make room for resize". Although, "in place update" + "lower priority pod eviction" seems to be overall way more complicated compared to a memory buffer.

[Swap](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2400-node-swap#long-running-applications-that-swap-out-startup-memory) is a step in the same direction as our proposal of the memory buffer. However, we have concerns about the [risks](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2400-node-swap#risks-and-mitigations) associated with introducing swap. For example, having swap available on a system reduces predictability. Swap's performance is worse than regular memory, sometimes by many orders of magnitude, which can cause unexpected performance regressions, etc.

**Comment by thockin** [comment]
This amounts to two-level scheduling - first place these "buffers" then place pods on nodes with the buffers.  It will require an extra level of cgroup to shift memory between specific pods (rather than all pods) and will impact all of the pod resource management code and accounting logic.

Ther would be a risk of scheduling mismatch, too - what if the buffer lands on a node and then no pods are able to land there?

TL;DR: This is a _huge_ change, I think.  While I don't want to rule it out entirely, it seems unlikely.

As @kannon92 said, swap seems like a better answer in the near term.  Another approach could be to just schedule a "balloon" pod which requests memory but doesn't use it.  That memory is available for all pods (not just these jobs) but this works today.

Another possible approach might be somthing with a tmpfs - not sure exactly how it would look, but something about pre-allocating memory that pods which know about it could access.

Usually, when I see buildfamr apps like this there are several tiers of jobs size (e.g. Small, Medium, Large).  Compiles first try on Small and if they fail they get moved to Medium, and repeat for large.  This means some jobs are a bit slower but resources are max efficient.

**Comment by syurevich** [comment]
@thockin thanks for your detailed answer and suggestions.

I think the most promising approach would be a "balloon" pod with a lower priority. This "balloon" pod can be evicted if there is not enough available memory on a node. It is also mentioned in [future-enhancements](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources#future-enhancements).

We will need to test if we can use such a "balloon" pod with a lower priority right away.

**Comment by thockin** [comment]
Another possibly interesting step towards this could be a per-namespace cgroup layer, so if you need to reclaim memory, you will apply pressure to your own pods first.  But there are pretty subtle implications of additional cgroups, so it's not to be entertained without a LOT of consideration :)
